xen/arch/arm/coproc/plat/gx6xxx.c

```diff
@@ -0,0 +1,303 @@
+#include <xen/irq.h>
+#include <xen/vmap.h>
+#include <asm/io.h>
+
+#include "../coproc.h"
+#include "common.h"
+
+#define DT_MATCH_GX6XXX DT_MATCH_COMPATIBLE("renesas,gsx")
+
+struct vgx6xxx_info
+{
+    /* This is the current IRQ status reported/updated to/from domains.
+     * Set on real IRQ from GPU.
+     */
+    uint32_t irq_status;
+
+};
+
+struct gx6xxx_info
+{
+    struct vcoproc_instance *curr;
+    uint32_t *reg_irq_status;
+    uint32_t *reg_irq_clear;
+};
+
+#if 0
+#define RGX_CR_MIPS_WRAPPER_IRQ_STATUS                (0x08A8U)
+#define RGX_CR_MIPS_WRAPPER_IRQ_STATUS_MASKFULL       (IMG_UINT64_C(0x0000000000000001))
+#define RGX_CR_MIPS_WRAPPER_IRQ_STATUS_EVENT_SHIFT    (0U)
+#define RGX_CR_MIPS_WRAPPER_IRQ_STATUS_EVENT_CLRMSK   (0XFFFFFFFEU)
+#define RGX_CR_MIPS_WRAPPER_IRQ_STATUS_EVENT_EN       (0X00000001U)
+
+#define RGX_CR_MIPS_WRAPPER_IRQ_CLEAR                 (0x08B0U)
+#define RGX_CR_MIPS_WRAPPER_IRQ_CLEAR_MASKFULL        (IMG_UINT64_C(0x0000000000000001))
+#define RGX_CR_MIPS_WRAPPER_IRQ_CLEAR_EVENT_SHIFT     (0U)
+#define RGX_CR_MIPS_WRAPPER_IRQ_CLEAR_EVENT_CLRMSK    (0XFFFFFFFEU)
+#define RGX_CR_MIPS_WRAPPER_IRQ_CLEAR_EVENT_EN        (0X00000001U)
+
+#define RGXFW_CR_IRQ_STATUS                           RGX_CR_MIPS_WRAPPER_IRQ_STATUS
+#define RGXFW_CR_IRQ_STATUS_EVENT_EN                  RGX_CR_MIPS_WRAPPER_IRQ_STATUS_EVENT_EN
+#define RGXFW_CR_IRQ_CLEAR                            RGX_CR_MIPS_WRAPPER_IRQ_CLEAR
+#define RGXFW_CR_IRQ_CLEAR_MASK                       RGX_CR_MIPS_WRAPPER_IRQ_CLEAR_EVENT_EN
+#else
+#define RGX_CR_META_SP_MSLVIRQSTATUS                  (0x0AC8U)
+#define RGX_CR_META_SP_MSLVIRQSTATUS_MASKFULL         (IMG_UINT64_C(0x000000000000000C))
+#define RGX_CR_META_SP_MSLVIRQSTATUS_TRIGVECT3_SHIFT  (3U)
+#define RGX_CR_META_SP_MSLVIRQSTATUS_TRIGVECT3_CLRMSK (0XFFFFFFF7U)
+#define RGX_CR_META_SP_MSLVIRQSTATUS_TRIGVECT3_EN     (0X00000008U)
+#define RGX_CR_META_SP_MSLVIRQSTATUS_TRIGVECT2_SHIFT  (2U)
+#define RGX_CR_META_SP_MSLVIRQSTATUS_TRIGVECT2_CLRMSK (0XFFFFFFFBU)
+#define RGX_CR_META_SP_MSLVIRQSTATUS_TRIGVECT2_EN     (0X00000004U)
+
+#define RGXFW_CR_IRQ_STATUS                           RGX_CR_META_SP_MSLVIRQSTATUS
+#define RGXFW_CR_IRQ_STATUS_EVENT_EN                  RGX_CR_META_SP_MSLVIRQSTATUS_TRIGVECT2_EN
+#define RGXFW_CR_IRQ_CLEAR                            RGX_CR_META_SP_MSLVIRQSTATUS
+#define RGXFW_CR_IRQ_CLEAR_MASK                       RGX_CR_META_SP_MSLVIRQSTATUS_TRIGVECT2_CLRMSK
+#endif
+
+static int gx6xxx_read(struct vcpu *v, mmio_info_t *info,
+                       register_t *r, void *priv)
+{
+    struct mmio *mmio = priv;
+    struct vcoproc_rw_context ctx;
+    static int start = 1;
+    uint32_t *vaddr;
+    unsigned long flags;
+
+    vcoproc_get_rw_context(v->domain, mmio, info, &ctx);
+    spin_lock_irqsave(&ctx.coproc->vcoprocs_lock, flags);
+    if (ctx.offset == RGXFW_CR_IRQ_STATUS) {
+        struct vgx6xxx_info *vinfo = (struct vgx6xxx_info *)ctx.vcoproc->priv;
+
```

Comment:
ctx.vcoproc might be NULL (inside vcoproc_get_rw_context() we call coproc_get_vcoproc(d, ctx->coproc) to fill it)